### ISSUE TITLE
Cross platform standard library

### ISSUE THREAD

**Issue by gradha** [rusthon/Rusthon#59]
The transpiled code seems to be highly backend dependant. Attempting something like a basic python slice already breaks c++ code:

```
def say_hi():
    print( 'hello world'[3:6])
==>
  File "./rusthon.py", line 539, in build
    pak = translate_to_cpp( pyjs, cached_json_files=cached_json )   ## pak contains: c_header and cpp_header
  File "<string>", line 13898, in translate_to_cpp
  File "<string>", line 1244, in visit
  File "<string>", line 13547, in visit_Module
  File "<string>", line 1244, in visit
  File "<string>", line 6341, in visit_FunctionDef
  File "<string>", line 12084, in _visit_function
  File "<string>", line 12212, in generate_generic_branches
  File "<string>", line 1244, in visit
  File "<string>", line 13723, in visit_Print
  File "<string>", line 1244, in visit
  File "<string>", line 10764, in visit_Subscript
__main__.GenerateSlice: {'upper': '6', 'lower': '3', 'slice': <_ast.Slice object at 0x102deabd0>, 'step': None, 'value': 'std::string("hello world")'}
```
Looks like to do some effective crossplatform program rusthon needs to reimplement a generic standard library which sometimes will map to the native standard library or implement some minimal code. Or maybe I have missed this support. Is it possible to write something like ``print(a.replace("h", "a"))`` and have it work on C++, Go, and Java?

**Comment by hartsantler** [comment]
I have created an issue here for fixing slice syntax.
https://github.com/rusthon/Rusthon/issues/60

Complete support for common python methods for strings and other builtins, is still a work in progress for all backends.  It requires each backend generates the right code that either maps to something native and inlines it, or calls some helper function from:
* https://github.com/rusthon/Rusthon/blob/master/src/cppheader.md
* https://github.com/rusthon/Rusthon/blob/master/src/runtime/cpp_builtins.py

**Comment by jcrubino** [comment]
This should be a priority and the standard tests can be built from this.
Where to start, Math, IO, Filesystem, OS?

**Comment by hartsantler** [comment]
i agree this should be a priority, and i encourage somebody to take on this large task.  the code could still use some clean ups and refactoring, and testing to ensure each backend works as closely as possible to standard python.  I can not help much on this because all those things are going to take awhile, and i'm overloaded with current tasks.

we should also try to reuse as much as possible from stdlib of RapydScript as possible for the JavaScript backend.

**Comment by techtonik** [comment]
This crosses needs with a PyPy project, where a lot of new languages written in RPython could benefit from such library with a known behaviour.

**Comment by jcrubino** [comment]
Great observation

On Tue, Aug 18, 2015 at 5:23 AM, anatoly techtonik <notifications@github.com

**Comment by hartsantler** [comment]
I have started to clean up some code, and getting the old regression tests for Go working again, soon all the tests in here will pass.
https://github.com/rusthon/Rusthon/tree/master/regtests/go

**Comment by wolfv** [comment]
I would be interested in helping to create a python standard library for the JS backend. E.g. starting with `math` and `random` and then looking further to what needs to be done. 

Would it make sense to create a repo under the rusthon umbrella for this purpose?

**Comment by hartsantler** [comment]
RapydScript might have a standard library that you can start with.

**Comment by techtonik** [comment]
Any cross-language / cross-platform library should start with some kind of language-independent tests of behavior.

**Comment by hartsantler** [comment]
on `visit_Import` we could trigger rapydscript to compile parts of its stdlib and insert the js into the output.

https://github.com/atsepkov/RapydScript/blob/master/src/lib/math.pyj
https://github.com/atsepkov/RapydScript/blob/master/src/lib/random.pyj

**Comment by hartsantler** [comment]
i have modified visit_Import in the javascript backend to call rapydscript in a subprocess and translate modules that is contained in `/lib/node_modules/rapydscript/src/lib/`

see this simple test:
https://github.com/rusthon/Rusthon/blob/master/regtests/stdlib/math_and_random.py
